src/ServiceControl/Operations/Watchdog.cs

```diff
@@ -0,0 +1,86 @@
+﻿namespace ServiceControl.Operations
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NServiceBus.Logging;
+
+    class Watchdog
+    {
+        Func<Task> ensureStopped;
+        Func<Task> ensureStarted;
+        Action<string> reportFailure;
+        Action clearFailure;
+        Task watchdog;
+        CancellationTokenSource shutdownTokenSource = new CancellationTokenSource();
+        TimeSpan timeToWaitBetweenStartupAttempts;
+        ILog log;
+        string processName;
+
+        public Watchdog(Func<Task> ensureStarted, Func<Task> ensureStopped, Action<string> reportFailure, Action clearFailure, TimeSpan timeToWaitBetweenStartupAttempts, ILog log, string processName)
+        {
+            this.ensureStopped = ensureStopped;
+            this.ensureStarted = ensureStarted;
+            this.reportFailure = reportFailure;
+            this.clearFailure = clearFailure;
+            this.timeToWaitBetweenStartupAttempts = timeToWaitBetweenStartupAttempts;
+            this.log = log;
+            this.processName = processName;
+        }
+
+        public Task OnFailure(string failure)
+        {
+            reportFailure(failure);
+            return ensureStopped();
+        }
+
+        public Task Start()
+        {
+            watchdog = Task.Run(async () =>
+            {
+                while (!shutdownTokenSource.IsCancellationRequested)
+                {
+                    try
+                    {
+                        await ensureStarted().ConfigureAwait(false);
+                        clearFailure();
+                    }
+                    catch (OperationCanceledException)
```

Comment:
In which case would this raise such an exception given that this API doesn't officially support cancellation?

Reply:
You're right. But now that I added the cancellation token to `ensureStarted` it makes sense to have a catch here, right?

Reply:
Then yes